github-actions/start-release/build_docs.py

```diff
@@ -0,0 +1,203 @@
+        for reserved_char in MARKDOWN_RESERVED_CHARACTERS:
+            text = text.replace(reserved_char, f"\\{reserved_char}")
+    return text
+
+
+def get_app_json(app_json_dir):
+    logging.info("Looking for app JSON in: %s", app_json_dir)
+    app_json_name = [f for f in os.listdir(app_json_dir)
+                     if f.endswith(".json")
+                     and "postman_collection" not in f][0]
+    json_file_path = Path(app_json_dir, app_json_name)
+    logging.info("Loading json: %s", app_json_name)
+    with open(json_file_path, "r") as json_file:
+        return json.load(json_file)
+
+
+def load_file(file_path):
+    try:
+        logging.info("Loading file: %s", file_path)
+        with open(file_path) as f:
+            return f.read()
+    except FileNotFoundError:
+        logging.warning("Couldn't find file: %s", file_path)
+    return None
+
+
+def render_template_to_file(connector_path, json_content):
+    env = Environment(
+        loader=FileSystemLoader(TEMPLATE_DIR),
+        extensions=[EscapeMDFilterVarsExtension],
+        autoescape=select_autoescape()
+    )
+    env.filters["generate_gh_fragment"] = generate_gh_fragment
+    env.filters["generate_action_heading_text"] = generate_action_heading_text
+    env.filters["escape_markdown"] = escape_markdown
+
+    logging.info("Rendering with template: %s from %s",
+                 TEMPLATE_NAME, TEMPLATE_DIR)
+    t = env.get_template(TEMPLATE_NAME)
+    rendered_content = t.render(connector=json_content)
+
+    output_readme_path = Path(connector_path, README_OUTPUT_NAME)
+    logging.info("Saved readme as %s", output_readme_path)
+    with open(output_readme_path, "w") as readme_file:
+        readme_file.write(rendered_content)
+
+    return rendered_content, output_readme_path
+
+
+def build_docs(connector_path, app_version=None):
+    input_readme_path = Path(connector_path, README_INPUT_NAME)
+
+    json_content = get_app_json(connector_path)
+    if app_version:
+        json_content["app_version"] = app_version
+
+    if md_content := load_file(input_readme_path):
+        template_path = Path(TEMPLATE_DIR, TEMPLATE_NAME)
+        with open(template_path) as template_file:
+            first_line_in_template = template_file.readline()
+
+        if first_line_in_template in md_content:
+            raise ValueError("Input readme contains auto-generated content!")
+
+    json_content["md_content"] = md_content
+
+    return render_template_to_file(connector_path, json_content)
+
+
+def manage_existing_markdown(connector_path):
+    md_path = Path(connector_path, README_INPUT_NAME)
+    author_notes_path = Path(connector_path, README_MD_ORIGINAL_NAME)
+
+    # Does the readme exist?
+    if not md_path.is_file():
+        return None, None
+
+    # Does notes.md already exist?
+    if author_notes_path.is_file():
+        logging.info("Removing existing auto-gen readme: %s", md_path)
+        md_path.unlink(missing_ok=True)
+        return None, None
```

Comment:
`notes.md` would only existing when running locally setting `override=False`?

Reply:
notes.md already exists for any app that already has a readme.md that was script-generated, *and* had a readme.md before conversion happened. Currently, this means there are a few apps already in GitHub with a readme.md.

Reply:
notes.md will be created in similar situations, when this script runs in the future. as written, the script doesn't need to be run locally for `overwrite=False` when `readme_html_to_markdown()` is called

Reply:
ah I see, we always take a backup of the previous `readme.md` as `notes.md`

do you think this still needs to happen when running in github, if we can also revert the given commit?

Reply:
just to be clear, are you suggesting we don't attempt to backup the original `readme.md` as `notes.md` before overwriting, and leave it to whoever ran the `start_release` job to notice if that happens?

Reply:
> ah I see, we always take a backup of the previous readme.md as notes.md
It doesn't always happen - it only happens if it seems to the script like the `readme.md` is not auto-generated.

Reply:
> just to be clear, are you suggesting we don't attempt to backup the original readme.md as notes.md before overwriting, and leave it to whoever ran the start_release job to notice if that happens?

The job writes the commit to `next` and creates a PR for `next` --> `main`, wouldn't the diff be shown in the PR?

Reply:
Yep, the change would be there in the diff for a human to notice or not notice. Then, if they notice, they could manually create another commit to preserve the overwritten content.

As it stands if they don't care about preserving the content that the script is currently backing-up automatically, they can notice and commit a change to remove `notes.md`.

The way the script is written is arguably a little safer and less subject to human error.

Reply:
sounds good, I was thinking that if a change *was* noticed then the commit could just be reverted to bring the readme back to it's original state, but having explicitly having a back up file could make people more likely to check for overwritten content